Separate submodule structure and status

gerhardol opened the pull request:
Part of #5769
(closes the issue with #6373 that may be dropped)

## Proposed changes
Submodule structure is separated from detailed status
Previously, the structure was always recalculated when status was updated
Also, the status is only calculated for the submodule trees where
git-status has detected changes.

This has simplified the code as well decreased the time before the
detailed status is displayed. The status was also flashing for several
seconds while updating (occurred regularly if any submodule had a change).

The normal use case with submodules is that most changes are done in the top project and that there are no or few changes in submodules.
#5777 (3.1.0) changed to only get the detailed status if there were any changes in submodules so the no-change scenario is handled already in master.
However, a change to any submodule would still require the status for all submodules.
Before #6372 this handling could hang the UI.
In addition to quicker update, the cpu load decreases with this PR. 

For a tree like below, previously was 5+8x6 git-diff commands required, with the update 1+2x6 (this is still an example where there are many dirty submodule trees).
This could of course be further improved by git-status on "dirty" levels, but the cost will likely be larger than the benefit (and complexity be increased).

![image](https://user-images.githubusercontent.com/6248932/55676420-4fb43800-58d5-11e9-96b3-1110f016e3bf.png)

Open:
 * Maybe update the test for submodule status handling (no such test existed before)
 * ~~AsyncLazy is not really needed, the status is serialized anyway~~

## Screenshots 
No change to the UI

## Test methodology
Test by making changes to submodules on several levels.
See #6357 for hints how to set this up
Check the Git Command log for expected commands

## Test environment(s)
See #6357 

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

RussKie commented on GitCommands/Submodules/SubmoduleStatusProvider.cs:
```diff
@@ -46,76 +60,98 @@ public void Dispose()
-
-        public void UpdateSubmodulesStatus(bool updateStatus, string workingDirectory, string noBranchText)
+        /// <inheritdoc />
+        public void UpdateSubmodulesStructure(GitModule gitModule, string noBranchText, bool updateStatus)
```
⚠️ `GitModule` should not be passed as an input parameter for a number of reasons.
1. It is a combination of state and behaviours, behavioural components should not be passed around.
2. Passing a concrete implementation creates a strong coupling, this reduces cohesion and negatively affects testability
3. Passing an object with a massive public surface in order to get a single property is also a bad practice.

Please rework the API in a way that does not require passing neither `IGitModule` nor `GitModule`.

gerhardol replied:
This will require some calls to Git for data that already exists.
OK to use an optimized copy from #6373?

gerhardol replied:
API changed (without  #6373 for now)

RussKie commented on GitCommands/Submodules/SubmoduleInfoResult.cs:
```diff
@@ -10,7 +10,7 @@ namespace GitCommands.Submodules
     {
         // Module that was used to gather submodule info
-        public IGitModule Module { get; internal set; }
+        public GitModule Module { get; internal set; }
```
Why?

gerhardol replied:
GetCurrentSubmoduleLocalPath() and SuperprojectModule() is not in IGitModule

gerhardol replied:
Passing currentModule around in addition to result (similar to how it was before)

RussKie replied:
Please add them to the interface

gerhardol replied:
GetCurrentSubmoduleLocalPath() can be added, but SuperprojectModule() is not trivial (you have discussed this with @amaiorano related to the sidepanel) as the return type need to be changed which will cause an avalanche of other additions to IGitModule. So that should be a separate refactoring.

gerhardol replied:
A proposal in https://github.com/gerhardol/gitextensions/tree/feature/i6359-submodule-selective-update-interfaces
However, requires the use of casts...

RussKie commented:
I'll pull it down and have a look in few days.

gerhardol commented:
Added a test for status handling. (No such test existed before, I had removed a test for some other functionality that no longer were needed.)
To do this, I removed AsyncLazy which no longer really was stricly needed after structure/status changes to be able to make some type of test.

gerhardol commented:
@jbialobr 
Do you have a possibility to review, especially the changes to Async?

gerhardol commented on UnitTests/GitCommandsTests/Submodules/SubmoduleStatusProviderTests.cs:
```diff
@@ -98,49 +98,31 @@ public void UpdateSubmoduleStatus_valid_result_for_second_nested_submodule()
                 changedFiles.Should().HaveCount(1);
-                _provider.HasChangedToNone(changedFiles).Should().BeFalse();
+                SubmoduleTestHelpers.UpdateSubmoduleStatusAndWaitForResult(_provider, currentModule, changedFiles);
+                ////result.OurSubmodules[0].Detailed.IsDirty.Should().BeTrue();
```
AsyncTestHelper.WaitForPendingOperations(); is not awaiting the status here, so the test fails (unless the debugger is connected).
I am guessing that is the test helper that is the problem.
Note: This was not tested before, not a new problem.

RussKie replied:
Your message looks cryptic...
It sounds like you're having a problem. Could you please explain further?

gerhardol replied:
> It sounds like you're having a problem.

The test is not working. (The reason this is still WIP)

The testhelper is not awaiting the result.
See below for a commit (in  https://github.com/gerhardol/gitextensions/pull/new/feature/i6359-submodule-selective-update-1) where the test works, but definitely worse API.

RussKie replied:
I had it enabled and it failed.
Then I force rebuilt and it worked...

gerhardol replied:
> I had it enabled and it failed.

Was this with this PR or the alternative branch?
I have not seen a solution for the testhelper or how to rewrite the tests

RussKie replied:
Maybe @jbialobr or @sharwell can help...

RussKie commented:
This branch.

On Wed, Apr 17, 2019, 6:22 PM Gerhard Olsson <notifications@github.com>
wrote:

> *@gerhardol* commented on this pull request.
> ------------------------------
>
> In UnitTests/GitCommandsTests/Submodules/SubmoduleStatusProviderTests.cs
> <https://github.com/gitextensions/gitextensions/pull/6445#discussion_r276129025>
> :
>
> >                  changedFiles.Should().HaveCount(1);
> -                _provider.HasChangedToNone(changedFiles).Should().BeFalse();
> +                SubmoduleTestHelpers.UpdateSubmoduleStatusAndWaitForResult(_provider, currentModule, changedFiles);
> +                ////result.OurSubmodules[0].Detailed.IsDirty.Should().BeTrue();
>
> I had it enabled and it failed.
>
> Was this with this PR or the alternative branch?
> I have not seen a solution for the testhelper or how to rewrite the tests
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/6445#discussion_r276129025>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEMyXvrhvb4XQb-SN0os52Kq4D6W-uMxks5vhtnDgaJpZM4cgeWr>
> .
>

gerhardol commented:
Added another test, also partly failing to document the expected behavior.
Removed WIP, the test (maybe after changes to the interface) can be in a follow up.

RussKie commented on GitUI/CommandsDialogs/FormBrowse.cs:
```diff
@@ -623,7 +616,7 @@ public void GoToRef(string refName, bool showNoRevisionMsg)
         }
 
-        #endregion
+#endregion
```
Please undo these changes

gerhardol commented:
Alternative tests in d3c699f using TestAccessor

gerhardol commented:
I want to squash...

RussKie commented:
Commendable desire! Don't deny yourself 😁

On Fri, Apr 19, 2019, 9:45 AM Gerhard Olsson <notifications@github.com>
wrote:

> I want to squash...
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/6445#issuecomment-484724716>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABBTEXWLPP62OOEHKBKJIKLPREBX7ANCNFSM4HEB4WVQ>
> .
>

gerhardol commented:
Rebased on top of new PR #6488 (to be able to enable one failed test)
Included the change to remove AsyncLazy, no longer separate.
